### ISSUE TITLE
File path highlighter breaks on `..` if pwd is a symlink

### ISSUE THREAD

**Issue by glittershark** [zsh-users/zsh-syntax-highlighting#176]
If my `pwd` is a symlink to another directory and I type `../somedirectory`, where `somedirectory` is a sibling of the symlink, the path will be highlighted as incorrect. However, if I type `../otherdirectory`, where `otherdirectory` is a sibling of the directory *that the symlink points to*, it will be highlighted as correct.

This also occurs for tab-completion in plain old zsh, so I'm not sure if it's really this library's fault

**Comment by danielshahaf** [comment]
For concreteness let's discuss a symlink `foo` that targets `bar/baz` where `baz` is a directory (a real directory, not a symlink).

In this case, after `cd foo`, both `cd ../bar` and `cd ../baz` work. However, most other commands really will accept only `../baz` and treat `../bar` as a non-existing path. Therefore, I think highlighting `../bar` as not-a-path is correct.

**Comment by glittershark** [comment]
`..` is `bar`, in your example...

**Comment by danielshahaf** [comment]
Yes, `foo/..` is `bar`.

**Comment by glittershark** [comment]
So of course `foo/../bar` won't be a path. I was referring to the following directory structure
```
foo
+---bar
+---baz
...
qux -> bar (PWD)
```

typing `../baz` will not highlight (but works just fine in all commands)

**Comment by danielshahaf** [comment]
Cannot reproduce:

```
% mkdir -p foo/bar foo/baz
% ln -s foo/bar qux 
% cd qux
% pwd 
/home/daniel/src/aux-zsh/z-sy-h/d/qux
% pwd -P
/home/daniel/src/aux-zsh/z-sy-h/d/foo/bar
% : ../baz 
```

The `../baz` on the last line is underlined.

Is this the directory structure you described?  I'm using revision 3dc574190071ef59054f6d693132410b37aa3ae4 and zsh-5.1-1-ga9df6aa.

**Comment by danielshahaf** [comment]
I'm trying this in `zsh -f` with no configuration except for `source zsh-syntax-highlighting.zsh`.

**Comment by glittershark** [comment]
Maybe this was fixed in 5.1.x - I'm running 5.0.8. I'll upgrade and see what happens

**Comment by glittershark** [comment]
Actually - this is the wrong way around. In your example `../baz` should not be underlined, but `../foo` should

**Comment by danielshahaf** [comment]
> Actually - this is the wrong way around. In your example ../baz should not be underlined, but ../foo should

No, highlighting `../baz` is correct, since `..` is a hard link to the parent of the physical directory, i.e., to `bar/.`.  Try running `stat ../foo` (errors out) and `stat ../baz` (succeeds).

**Comment by glittershark** [comment]
it's not a hard link, it's a soft (symbolic) link

**Comment by danielshahaf** [comment]
`qux` is a soft link, but the `..` dirent is a hard link:

```
% zmodload zsh/stat
% readlink qux
foo/bar
% zstat +inode qux/..
343537
% zstat +inode foo
343537
```

**Comment by danielshahaf** [comment]
@glittershark I think you are simply confused and the code is correct: what the code underlines, `stat` succeeds on, and what the code doesn't underline, `stat` errors out on (with the error `No such file or directory`).

**Comment by glittershark** [comment]
I understand now. I still think the (technically correct) functionality may be functionally unintuitive, because `cd ..` will `cd` to `basename $PWD` rather than `readlink ..`

**Comment by danielshahaf** [comment]
> I understand now. I still think the (technically correct) functionality may be functionally unintuitive, because cd .. will cd to the parent path of $PWD rather than readlink ..

Yes, `cd` doesn't follow the normal rules. For example, both `cd ../foo` and `cd ../baz` work, even though they require different and inconsistent interpretations of `..`. (This is even with the `CHASE_DOTS` option unset, as it is by default.)

Bottom line, since `cd ../foo` works, it would make sense to highlight `../foo` as a path _in that specific context_, i.e., for the `cd` builtin (but not for everything). Let's leave this issue open to track that enhancement.
